### PR TITLE
Add protocol/ws module with immutable flyweights

### PR DESCRIPTION
**Pull request by cmebarrow**
This is for story [#106 nuklei.ws framing](https://github.com/kaazing/roadmap/issues/106)

New submodule protocol/ws is added, along with Flyweight subclasses. Also adds Utf8Util to core for UTF8 validation.

### CODE REVIEW

File: core/src/main/java/org/kaazing/nuklei/ErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.kaazing.nuklei;
+
+@FunctionalInterface
+public interface ErrorHandler
+{
+
+    void handleError(String message);
+
+}
```
Comment (dpwspoon) [inline core/src/main/java/org/kaazing/nuklei/ErrorHandler.java]: Formatting: Files should end with new line

File: protocol/ws/pom.xml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                            <goal>start</goal>
+                            <goal>stop</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <configuration>
```
Comment (dpwspoon) [inline protocol/ws/pom.xml]: I believe this configuration is inherited, thus doesn't need to be specified

File: protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/Data.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        this.maxWsMessageSize = maxWsMessageSize;
+    }
+
+    @Override
+    public Payload getPayload()
+    {
+        Payload payload = super.getPayload();
+        if (getOpCode() == OpCode.TEXT)
+        {
+            Utf8Util.validateUTF8(payload.buffer(), payload.offset(), getLength(), (message) ->
```
Comment (dpwspoon) [inline protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/Data.java]: Do we need to validate it here?

File: protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/OpCode.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        case PING:
+            return 0x09;
+        case PONG:
+            return 0x0A;
+        default:
+            throw new ProtocolException(format("Unrecognised OpCode %s", value));
+        }
+    };
+
+}
```
Comment (dpwspoon) [inline protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/OpCode.java]: newline

**Comment by dpwspoon** [comment]
Looks good to me

File: protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/Data.java (hunk 3 above)
Comment (cmebarrow) [inline protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/Data.java]: RFC 6455 mandates using valid UTF8 in text frames. Attempting to input invalid UTF8 is a common attack vector, so I think we have to protect against it. There may be cases where we don't need to valid (e.g. proxy service connecting to another gateway service where we do validate) but in that case the service code can just forward the packet unaltered without calling getPayload.

File: protocol/ws/pom.xml (hunk 2 above)
Comment (cmebarrow) [inline protocol/ws/pom.xml]: Thanks, I have removed the plugin configurations so they are inherited.

File: core/src/main/java/org/kaazing/nuklei/ErrorHandler.java (hunk 1 above)
Comment (cmebarrow) [inline core/src/main/java/org/kaazing/nuklei/ErrorHandler.java]: Fixed

File: protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/OpCode.java (hunk 4 above)
Comment (cmebarrow) [inline protocol/ws/src/main/java/org/kaazing/nuklei/protocol/ws/codec/OpCode.java]: fixed

**Pull request merged**